src/pip/_internal/utils/misc.py

```diff
@@ -510,6 +523,12 @@ def unzip_file(filename, location, flatten=True):
                 fn = split_leading_dir(name)[1]
             fn = os.path.join(location, fn)
             dir = os.path.dirname(fn)
+            # https://github.com/pypa/pip/issues/3907
+            if not is_within_directory(location, str(fn)):
+                raise InstallationError(
+                    'A file (%s) is trying to install outside location (%s)' %
```

Comment:
I think it would also be helpful to include the name of the zip file itself that it's failing on (not just the name of the file inside the zipfile). I also think "target directory" would be clearer than "location."